lib/symboldatabase.cpp

```diff
@@ -2474,6 +2476,20 @@ const Token * Function::constructorMemberInitialization() const
     return nullptr;
 }
 
+bool Function::isBodyEmpty() const
+{
+    if (!hasBody())
+    {
+        return true;
```

Comment:
In general `hasBody` will return false when the implementation is in another TU. So there is probably a body somewhere we just can't see it. It seems unsafe to guess that the body is empty.

Reply:
Does any other way exist to check class for constructor side-effects? I thought out only this.

Reply:
The best way would be to use the whole program analysis;
 * Create a class `CheckClass::FileInfo` that derives from `Check::FileInfo`.
 * Add a function `CheckClass::getFileInfo` that overrides the `Check::getFileInfo` method.. and in that determine if all public class in the TU have side effects or not. Also determine if there is "possible bad usage" of classes.
 * Add a function `CheckClass::analyseWholeProgram` that overloads `Check::analyseWholeProgram`.. there you can check if "possible bad usage" is actually "bad".

Reply:
For information.. if you want to perform the whole program analysis in a separate PR that is fine for me. You can just ensure in this PR that there is not a false positive. And in a later PR you can fix the false negatives with whole program analysis.

Reply:
I'll try to make whole analysis as a separate change.
Please check [this commit](https://github.com/danmar/cppcheck/pull/2692/commits/7bde68a0196e0f1c0ca58d2c74abfa088c155ddd). I'm not sure is it the test you are talking about.